Content.Server/GameObjects/Components/Kitchen/ReagentGrinderComponent.cs

```diff
@@ -0,0 +1,367 @@
+
+            if (UserInterface != null)
+            {
+                UserInterface.OnReceiveMessage += UserInterfaceOnReceiveMessage;
+            }
+
+            _audioSystem = EntitySystem.Get<AudioSystem>();
+        }
+
+        private void UserInterfaceOnReceiveMessage(ServerBoundUserInterfaceMessage message)
+        {
+            if(!Powered || _busy)
+            {
+                return;
+            }
+
+            switch(message.Message)
+            {
+                case ReagentGrinderGrindStartMessage msg:
+                    ClickSound();
+                    DoWork(user:message.Session.AttachedEntity!,isJuiceIntent:false);
+                    break;
+
+                case ReagentGrinderJuiceStartMessage msg:
+                    ClickSound();
+                    DoWork(user: message.Session.AttachedEntity!,isJuiceIntent:true);
+                    break;
+
+                case ReagentGrinderEjectChamberAllMessage msg:
+                    if(!ChamberEmpty)
+                    {
+                        ClickSound();
+                        for (var i = _chamber.ContainedEntities.Count - 1; i >= 0; i--)
+                        {
+                            EjectSolid(_chamber.ContainedEntities.ElementAt(i).Uid);
+                        }
+                    }
+                    break;
+
+                case ReagentGrinderEjectChamberContentMessage msg:
+                    if (!ChamberEmpty)
+                    {
+                        EjectSolid(msg.EntityID);
+                        ClickSound();
+                        _dirty = true;
+                    }
+                    break;
+
+                case ReagentGrinderEjectBeakerMessage msg:
+                    ClickSound();
+                    EjectBeaker(message.Session.AttachedEntity!);
+                    //EjectBeaker will dirty the UI for us, we don't have to do it explicitly here.
+                    break;
+            }
+        }
+        private void ClickSound()
+        {
+            _audioSystem.PlayFromEntity("/Audio/Machines/machine_switch.ogg", Owner, AudioParams.Default.WithVolume(-2f));
+        }
+        private void SetAppearance(ReagentGrinderVisualState state)
+        {
+            if (Owner.TryGetComponent(out AppearanceComponent? appearance))
+            {
+                appearance.SetData(PowerDeviceVisuals.VisualState, state);
+            }
+        }
+
+        public void OnUpdate()
+        {
+            if(_dirty)
+            {
+                UserInterface?.SetState(new ReagentGrinderInterfaceState
+                (
+                    _busy,
+                    HasBeaker,
+                    _chamber.ContainedEntities.Select(item => item.Uid).ToArray(),
+                    //Remember the beaker can be null!
+                    _heldBeaker?.Solution.Contents.ToArray()
+                ));
+                _dirty = false;
+            }
```

Comment:
If _dirty can be set more than once in a tick then IMO just have an event to add it to the system, then the system runs through any dirty reagentgrinders in a tick when necessary (rather than ticking over every single one). It's not a huge deal given there's probably not a lot of them but it'll be a benefit.

If _dirty can only be set once then just have it set the state directly given that ServerUserInterfaceComponent (the manager for BoundUserInterface) already handles this.

Reply:
tldr